src/libraries/slngen.proj

```diff
@@ -0,0 +1,85 @@
+    <NuGetConfigTemplatePath>$(RepositoryEngineeringDir)slngen.nuget.config</NuGetConfigTemplatePath>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <SolutionFile Include="$(MSBuildThisFileDirectory)*\*.sln" />
+    <SourceProject Include="$(MSBuildThisFileDirectory)*\src\*.*proj" />
+  </ItemGroup>
+
+  <Target Name="UpdateSolutionFile"
+          AfterTargets="Build"
+          DependsOnTargets="GetSolutionFiles"
+          Inputs="@(SolutionFile)"
+          Outputs="%(ProjFilePath)">
+    <!-- Copy template files, required by slngen -->
+    <Copy SourceFiles="$(ProjTemplatePath)"
+          DestinationFiles="%(SolutionFile.ProjFilePath)" />
+    
+    <!-- Invoke slngen -->
+    <Exec Command="dotnet slngen &quot;%(SolutionFile.ProjFilePath) -p SlnGenMainProject=%(SolutionFile.Filename) --launch false --nologo&quot;" />
+
+    <!-- Delete temporary template file -->
+    <Delete Files="%(SolutionFile.ProjFilePath)" />
+  </Target>
+
+  <Target Name="AddNuGetConfigForRestoreToSolution"
+          AfterTargets="UpdateSolutionFile"
+          DependsOnTargets="GetSolutionFiles;GetSourceProjectWithDuplicateTfm"
+          Inputs="@(SolutionFile)"
+          Outputs="%(NuGetConfigFilePath)">
+    <!-- Delete existing NuGet.config files -->
+    <Delete Files="%(SolutionFile.NuGetConfigFilePath)" />
+    
+    <!-- Check if solution file contains a source project with a duplicate tfm -->
+    <PropertyGroup>
+      <SolutionFileContent>$([System.IO.File]::ReadAllText('%(SolutionFile.Identity)'))</SolutionFileContent>
+      <SolutionFileContainsSourceProjectWithDuplicateTfm />
+      <SolutionFileContainsSourceProjectWithDuplicateTfm Condition="$(SolutionFileContent.Contains('src\%(SourceProjectWithDuplicateTfm.Identity)'))">true</SolutionFileContainsSourceProjectWithDuplicateTfm>
+    </PropertyGroup>
+    
+    <!-- Copy NuGet config template file -->
+    <Copy SourceFiles="$(NuGetConfigTemplatePath)"
+          DestinationFiles="%(SolutionFile.NuGetConfigFilePath)"
+          Condition="'$(SolutionFileContainsSourceProjectWithDuplicateTfm)' == 'true'" />
+    
+    <Message Text="Copying NuGet.config template file for %(SolutionFile.Filename)"
+             Importance="high"
+             Condition="'$(SolutionFileContainsSourceProjectWithDuplicateTfm)' == 'true'" />
+  </Target>
+
+  <Target Name="GetSolutionFiles">
+    <ItemGroup>
+      <!-- Add attributes that require a separate item mutation. -->
+      <SolutionFile ProjFilePath="%(RelativeDir)%(Filename).proj"
+                    NuGetConfigFilePath="%(RelativeDir)NuGet.config" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="GetSourceProjectWithDuplicateTfm"
+          Inputs="@(SourceProject)"
+          Outputs="%(Identity)">
+    <!-- Check if project file contains duplicate tfms -->
+    <PropertyGroup>
+      <ProjectFileContent>$([System.IO.File]::ReadAllText('%(SourceProject.Identity)'))</ProjectFileContent>
```

Comment:
Why are we doing this instead of just calling a target in the project?  Is it for perf?  Is there a way to do a faster load of a project (eg: skipping imports or something) to improve perf, but still net MSBuild do the parsing/eval of the project?

Reply:
There isn't a target in the SDK currently that returns the TargetFrameworks string in its raw form which we need in this case. I also wouldn't want to call into msbuild to eval/parse the project as it just isn't worth it.

Reply:
Are you concerned with folks using properties in the TargetFrameworks string or using something that fools these regexs?

A couple small  comments on the method, if you keep the regex approach, avoid raising the entire project content to a property, instead feed it directly into the regex.  Also, it seems like your regex is already matching the TargetFrameworks tags, so you should be able to avoid using string.Replace calls by setting property to the captured content.

Reply:
> Also, it seems like your regex is already matching the TargetFrameworks tags, so you should be able to avoid using string.Replace calls by setting property to the captured content.

Unfortunately using Regex.Match as a property function doesn't allow to specify an inner capture group. It only returns the full capture which includes the boundaries.

Reply:
> Are you concerned with folks using properties in the TargetFrameworks string or using something that fools these regexs?

Not at all, I tested it with properties as well. It works with the inlined values or with the markers like `$(NetCoreAppCurrent)`. I'm not concerned about derived properties as we don't have those today and we probably add ones for anything older than net5.0.

Reply:
One more thing. I spoke with Rainer about this and there currently isn't a way from within msbuild to do this without writing a task which I wanted to avoid. There's a proposal though: https://github.com/dotnet/msbuild/issues/3911.

Reply:
OK, really last comment. The TargetFramework.Sdk today mutates the TargetFrameworks property which is yet another reason why I don't want to depend on the evaluated values. As I see this as a point in time probably I'm ok with this solution.

Reply:
That's fair. You might be able to use Regex replace instead of match in order to use a capture group.￼

Reply:
That's possible but if you don't mind I would just leave as is as CI is already green and this would only be code clean-up (removing two lines).

Reply:
Sure, this is less blocking since this is just a manually run tool